mmedit/models/restorers/ttsr.py

```diff
@@ -0,0 +1,259 @@
+            log_vars=log_vars,
+            num_samples=len(gt.data),
+            results=dict(
+                lq=lq.cpu(), gt=gt.cpu(), ref=ref.cpu(), output=pred.cpu()))
+
+        pred = None
+        loss = None
+
+        return outputs
+
+    def forward_test(self,
+                     lq,
+                     lq_up,
+                     ref,
+                     ref_downup,
+                     gt=None,
+                     meta=None,
+                     save_image=False,
+                     save_path=None,
+                     iteration=None):
+        """Testing forward function.
+
+        Args:
+            lq (Tensor): LQ image
+            gt (Tensor): GT image
+            lq_up (Tensor): Upsampled LQ image
+            ref (Tensor): Reference image
+            ref_downup (Tensor): Image generated by sequentially applying
+                bicubic down-sampling and up-sampling on reference image
+            meta (list[dict]): Meta data, such as path of GT file.
+                Default: None.
+            save_image (bool): Whether to save image. Default: False.
+            save_path (str): Path to save image. Default: None.
+            iteration (int): Iteration for the saving image name.
+                Default: None.
+
+        Returns:
+            dict: Output results, which contain either key(s)
+                1. 'eval_result'.
+                2. 'lq', 'pred'.
+                3. 'lq', 'pred', 'gt'.
+        """
+
+        # generator
+        with torch.no_grad():
+            pred = self.forward_dummy(
+                lq=lq, lq_up=lq_up, ref=ref, ref_downup=ref_downup)
+
+        pred = (pred + 1.) / 2.
+        if gt is not None:
+            gt = (gt + 1.) / 2.
+
+        if self.test_cfg is not None and self.test_cfg.get('metrics', None):
+            assert gt is not None, (
+                'evaluation with metrics must have gt images.')
+            results = dict(eval_result=self.evaluate(pred, gt))
+        else:
+            results = dict(lq=lq.cpu(), output=pred.cpu())
+            if gt is not None:
+                results['gt'] = gt.cpu()
+
+        # save image
+        if save_image:
+            if 'gt_path' in meta[0]:
+                the_path = meta[0]['gt_path']
+            else:
+                the_path = meta[0]['lq_path']
+            folder_name = osp.splitext(osp.basename(the_path))[0]
+            if isinstance(iteration, numbers.Number):
+                save_path = osp.join(save_path, folder_name,
+                                     f'{folder_name}-{iteration + 1:06d}.png')
+            elif iteration is None:
+                save_path = osp.join(save_path, f'{folder_name}.png')
+            else:
+                raise ValueError('iteration should be number or None, '
+                                 f'but got {type(iteration)}')
+            mmcv.imwrite(tensor2img(pred), save_path)
+
+        return results
+
+    def val_step(self, data_batch, **kwargs):
```

Comment:
This can be inherited from `BasicRestorer`.

Reply:
Sure, I will remove the function later.